File: leetcode/python/0136_single_number.py
class Solution(object):
    def singleNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        nums.sort()
        # 1 1 2 2 4
        # 1 2 2
        # 1 1 2 3 3
        if len(nums) == 1: return nums[0]

        i, j, res = 0, 1, 0

        while j <= len(nums):
            if j == len(nums): return nums[-1]

            if nums[i] != nums[j]: return nums[i]

            i += 2
            j += 2

        return res

File: leetcode/java/0136_single_number.java
class Solution {
    public int singleNumber(int[] nums) {
        // sort nums
        // if nums.length == 1: return nums[1]
        // loop through nums using i (starting 0), j (starting 1)
        // if (j == nums.length) return nums[i]
        // if (nums[i] != nums[j]) return nums[i]
        // i++; j++;

        Arrays.sort(nums);
        if (nums.length == 1) return nums[0];
        int i = 0, j = 1, res = 0;

        while (j <= nums.length) {
            if (j == nums.length) {
                res = nums[i];
                break;
            }
            if (nums[i] != nums[j]) {
                res = nums[i];
                break;
            }

            i+=2;
            j+=2;
        }

        return res;
    }
}